src/main/scala/strawman/collection/immutable/Vector.scala

```diff
@@ -0,0 +1,1140 @@
+package strawman
+package collection
+package immutable
+
+import strawman.collection.mutable.{Builder, ReusableBuilder}
+
+import scala.annotation.unchecked.uncheckedVariance
+import scala.{AnyRef, Array, Boolean, IllegalArgumentException, IndexOutOfBoundsException, `inline`, Int, math, NoSuchElementException, Nothing, Serializable, SerialVersionUID, Unit, UnsupportedOperationException}
+import scala.Predef.intWrapper
+
+/** Companion object to the Vector class
+ */
+object Vector extends IterableFactory[Vector] {
+
+  def empty[A]: Vector[A] = NIL
+
+  def fromIterable[E](it: collection.Iterable[E]): Vector[E] =
+    it match {
+      case v: Vector[E] => v
+      case _            => NIL ++ it
+    }
+
+  def newBuilder[A](): Builder[A, Vector[A]] = new VectorBuilder[A]
+
+  private[immutable] val NIL = new Vector[Nothing](0, 0, 0)
+
+  // Constants governing concat strategy for performance
+  private final val Log2ConcatFaster = 5
+  private final val TinyAppendFaster = 2
+}
+
+// in principle, most members should be private. however, access privileges must
+// be carefully chosen to not prevent method inlining
+
+/** Vector is a general-purpose, immutable data structure.  It provides random access and updates
+ *  in effectively constant time, as well as very fast append and prepend.  Because vectors strike
+ *  a good balance between fast random selections and fast random functional updates, they are
+ *  currently the default implementation of immutable indexed sequences.  It is backed by a little
+ *  endian bit-mapped vector trie with a branching factor of 32.  Locality is very good, but not
+ *  contiguous, which is good for very large sequences.
+ *
+ *  $usesMutableState
+ *
+ *  @see [[http://docs.scala-lang.org/overviews/collections/concrete-immutable-collection-classes.html#vectors "Scala's Collection Library overview"]]
+ *  section on `Vectors` for more information.
+ *
+ *  @tparam A the element type
+ *
+ *  @define Coll `Vector`
+ *  @define coll vector
+ *  @define thatinfo the class of the returned collection. In the standard library configuration,
+ *    `That` is always `Vector[B]` because an implicit of type `CanBuildFrom[Vector, B, That]`
+ *    is defined in object `Vector`.
+ *  @define bfinfo an implicit value of class `CanBuildFrom` which determines the
+ *    result class `That` from the current representation type `Repr`
+ *    and the new element type `B`. This is usually the `canBuildFrom` value
+ *    defined in object `Vector`.
+ *  @define orderDependent
+ *  @define orderDependentFold
+ *  @define mayNotTerminateInf
+ *  @define willNotTerminateInf
+ */
+@SerialVersionUID(-1334388273712300479L)
+final class Vector[+A] private[immutable] (private[collection] val startIndex: Int, private[collection] val endIndex: Int, focus: Int)
+  extends IndexedSeq[A]
+    with IndexedSeqOps[A, Vector, Vector[A]]
+    with VectorPointer[A @uncheckedVariance]
+    with Serializable { self =>
```

Comment:
I essentially copy-pasted the current implementation and adapted it to the new design when needed.